Update manage-search-schema doc with new user query filters (#509)
sharetribe/flex-docs: src/docs/flex-cli/manage-search-schemas-with-flex-cli/index.md

[Getting started with Flex CLI](/flex-cli/getting-started-with-flex-cli/).

In this tutorial, we will add data schemas for the `category` and
`amenities` public data fields for listings. And then `age` protected data
for users. New marketplaces don't have any schemas in the backend by default
since the needs of marketplaces vary.

## Extended data types and schema scopes

There are various kinds of extended data. Search schema support differs depending on whether you
use Marketplace API or Integration API, and on whether the `schema-for` value is `listing` or `user`.
The following table summarizes the supported search schema scopes.

| schema-for | Supported scopes                 |
|------------|----------------------------------|
| listing    | public, meta, keyword            |

> | user       | public, meta, private, protected |

Author: 👍  good point!

> <extrainfo title="Currently there's no support for extended data for users in Marketplace API, so the search schema is only available for users in Integration API">

Author: 👍

## Extended data types and schema scopes

There are various kinds of extended data. Search schema support differs depending on whether you
use Marketplace API or Integration API, and on whether the `schema-for` value is `listing` or `user`.
The following table summarizes the supported search schema scopes.

> |          | Marketplace API                | Integration API                  |

Author: @bladealslayer You are right but do you think there won't be any users who try to use `user` data schema search using Marketplace API if we adopt your version?

If we don't explain the differences in Marketplace and Integration API, I think there might be some confusion until Marketplace API catches up. If you are sure of this, ofc, then I will change this.

> |          | Marketplace API                | Integration API                  |

Reviewer: Would it make sense to add a paragraph or an `<extrainfo>` block that explains that currently there's no endpoint to query users in Marketplace API, so the search schema is in practice only available for Integration API? I tend to agree with Won that it makes sense to mention the practical difference between the APIs somewhere, even though technically the same schemas are supported under the hood.

> |          | Marketplace API                | Integration API                  |

Author: @SariSaar I took your advice :) let me know if this looks good.

@bladealslayer too

| -------- | ------------------------------ | -------------------------------- |
| listing  | public, meta, keyword          | public, meta, keyword            |
| user     | N/A                            | public, meta, private, protected |

Extended data are editable in Console by the operator, but only public data and 
metadata can be seen by other users. For this reason, search schemas can be scoped 

> either to `public` or `metadata`. To see more details about extended data, see the

Reviewer: The table implies that `user` search schemas can be scoped to `private` and `protected` as well, so this sentence should be reformulated somehow.

Author: Yes, true. Didn't change here much as I didn't use console for extended data. I will verify the situation.

Reviewer: The way I read this, the paragraph talks about search schemas in general, not separating them between `listing` and `user`. I'd suggest making the distinction explicit - something along the following lines:

"All types of extended data are editable in Console by the operator, but only public data and 
metadata can be seen by other marketplace users. For this reason, listing search schemas can be scoped  either to `public` or `metadata`. User search schemas are currently only available through an Integration API endpoint, so they can also be scoped to `private` and `protected`."

Author: @bladealslayer 
I more or less adopted @SariSaar 's suggestion but I took out `For this reason` because there is no strong causal relationship. `listing's `private`, `protected` data schema should be enabled in integration API as well`. And I used the word `Currently` to imply situation could change.

## Extended data types and schema scopes

There are various kinds of extended data. Search schema support differs depending on whether you

> use Marketplace API or Integration API and also on whether the `schema-for` value is `listing` or `user`. The below table summarizes what scopes of search schema is supported based on different factors.

Reviewer: ```suggestion
use Marketplace API or Integration API and also on whether the `schema-for` value is `listing` or `user`. The following table summarizes the supported search schema scopes.
```

|          | Marketplace API                | Integration API                  |
| -------- | ------------------------------ | -------------------------------- |

> | listing  | public, meta, private, keyword | public, meta, keyword (private, protected coming in the future) |

Reviewer: TODO: If Marketplace API listings query allows private search schemas, API reference requires an update as well, since it only lists pub_* and meta_* as possible query params for [Marketplace API listing query](https://www.sharetribe.com/api-reference/marketplace.html#query-listings).

Author: Oh I think this has been my mistake :) will fix it.

Reviewer 2: I wouldn't mention any future plans here, as they are uncertain and there's no point to over-promise something we might not do anytime soon.

Author: @bladealslayer 

I will delete the mention about future plans. But I know we have discussed this could come in future that's why I mentioned them here.

"If we know" that they are coming, I imagined a few good effects of mentioning planned features. 

* Existing customers can plan what to do with our future features. And "if they are planning to quit our subscription because we are missing X, Y and Z", we can reassure them they are coming soon they just need to wait some time.
* Even though some potential customers know that they are missing some features right now, we can let them know we have plans for them, and this might convince them to start using Flex now.
* If they read our docs, it can prevent questions like "Are you planning to add X, Y and Z features?" (well we can still get questions like "When are you planning to do X, Y and Z"...
* There might be more but more or less based on the same logic...

I am totally fine with not mentioning any planned features and deleting them now. This is just food for thoughts and we wanna consider the positive side of the trade offs as well.

Reviewer 2: Yeah, indeed. However, if we don't really have concrete plans on building those, people may have the wrong expectation and think we've promised those features but aren't delivering them.

| user     | N/A                            | public, meta, private, protected |

> For extended data exposed via Marketplace API, they are editable in Console by

Reviewer 2: This isn't exactly true. All extended data is editable in Console, regardless of whether it's exposed in the Marketplace API or not.

Author: Fixing it. `but only public data and metadata can be seen by other users.` this part may also need some advice.

| long       | one         | `distanceToLake: 30`                                      | `pub_distanceToLake=5,40`                                                          |
| text       | one         | `stoveDescription: "Modern and powerful electric stove."` | `keywords=powerful%20modern`                                                       |

Note that the scope in the examples above is `public`. Please use correct prefix

> depending on the prefix (`meta_` for metadata, `priv_` for private data,

Reviewer 3: ```suggestion
depending on the scope of the data (`meta_` for metadata, `priv_` for private data,
```

flex-cli search set --key amenities --type multi-enum --scope public -m my-test-marketplace
```

Note that we don't need to supply `--schema-for` option when we add schema for `listing` as `listing` is the default.

> We can also add a search schema for `userProfile` with `long` type.

Reviewer 2: The current section, under which this text is, is written more like a concrete tutorial for adding listing search schema. As such, this text doesn't make sense there, because the operator is not interested in adding userProfile schema at that moment. This could perhaps be a new heading "Adding user search schema".

Note that the opening paragraphs in this article also present it as tutorial for managing listing search schema. So some rephrasing is needed.

Author: * opening paragraphs changed
* new heading "Adding user search schema" added
* existing heading renamed "Adding schemas" -> "Adding listing search schemas"

> And we can also add a search schema for `userProfile` with `long` type.

Reviewer: ```suggestion
We can also add a search schema for `userProfile` with `long` type.
```

```shell
flex-cli search set --schema-for userProfile --key age --type long --scope protected -m my-test-marketplace
```

> We should now see the details of those schemas:

Reviewer: If the `userProfile` schema is added as in the example, should it show in the schema details i.e. should the image `search-schemas-added.png` be replaced with one that shows the `userProfile`  schema as well?

Author: Good catch. I should first deploy some changes to staging so that I can take screenshot from my terminal. I will keep in mind.

Author: The image is replaced.

flex-cli search set --key amenities --type multi-enum --scope public -m my-test-marketplace
```

> Note that we don't need to supply `--schema-for` option when we add schema for `listing` as `listing` is the default.

Reviewer 2: I remember some time ago a rule of thumb to avoid "we" and "you" distinctions in our docs or communication. Something about it creating a polarity in the reader's mind. I personally have been avoiding that, but not sure it's always that easy. What do you think, @SariSaar?

Author: That's an interesting point. Should I change it to:

Note that `--schema-for` option is not needed when adding schema for `listing` as `listing` is the default.

?

> ## Adding user search schema

Reviewer 2: Maybe it would be nice to add a bit of intro on how user profile schema is useful:

"
User profile search schema can be useful, if you have an Integration API application that needs to query different sets of users, depending on some value in the user profile's extended data. For instance, if users have `age` attribute stored in their protected data, you can use the [/users/query endpoint in the Integration API](/references/extended-data/) to find users of a certain age range.

Search schema for user profiles can be added as follows:
"

Author: Nice suggestion, added the intro as suggested. :)

We can also add a search schema for `userProfile` with `long` type.

```shell
flex-cli search set --schema-for userProfile --key age --type long --scope protected -m my-test-marketplace
```

We should now see the details of those schemas:

> ![Search schemas for listing category, amenities and user age](./search-schemas-added.png)

Reviewer 2: @wontheone1 the colors in the screenshot are a bit hard to separate - especially the bolded black ones. Can you adjust your terminal to have solid background and maybe different lighter color for the bolds? See, e.g. the old images - there the table column titles are visible better.

Reviewer 2: BTW, you can see the resulting docs site by clicking the "Details" link on the Netlify automatic check that appears on the PR. The last one on the list " @netlify
netlify/sharetribe-flex-docs-site/deploy-preview — Deploy Preview ready! " is the one for the live preview.

That way you can more easily spot formatting issues and see the actual rendered articles.

## Adding user search schema

> We can also add a search schema for `userProfile` with `long` type.

Author: Also avoided using `We` here.

User profile search schema can be useful, if you have an Integration API application that
needs to query different sets of users, depending on some value in the user profile's extended data.
For instance, if users have age attribute stored in their protected data, you can use the
[/users/query endpoint in the Integration API](https://www.sharetribe.com/docs/references/extended-data/)
to find users of a certain age range.

Search schema for user profiles can be added as follows:

```shell
flex-cli search set --schema-for userProfile --key age --type long --scope protected -m my-test-marketplace
```

> We should now see the details of those schemas:

Author: Also avoided using `We` here.
